File: src/integrator/LangevinBarostat.cpp
/*
  Copyright (C) 2012,2013
      Max Planck Institute for Polymer Research
  Copyright (C) 2008,2009,2010,2011
      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
  
  This file is part of ESPResSo++.
  
  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
*/

#include "python.hpp"
#include "LangevinBarostat.hpp"

#include "types.hpp"
#include "System.hpp"
#include "storage/Storage.hpp"
#include "iterator/CellListIterator.hpp"

#include "interaction/Interaction.hpp"
#include "esutil/RNG.hpp"
#include "esutil/Error.hpp"

#include "bc/BC.hpp"

#include "mpi.hpp"

namespace espressopp {

  using namespace std;
  
  using namespace iterator;
  using namespace interaction;

  namespace integrator {

    LOG4ESPP_LOGGER(LangevinBarostat::theLogger, "LangevinBarostat");

    // TODO The ensemble is very sensitive to parameters. The manual how to choose the
    // parameters for simple system should be written.
    LangevinBarostat::LangevinBarostat(shared_ptr<System> _system,
                                       shared_ptr<esutil::RNG> _rng,
                                       real _temperature) : Extension(_system), 
                                                            rng(_rng), 
                                                            desiredTemperature(_temperature){
      // external parameters
      gammaP = 0.0;
      mass = 0.0;
      externalPressure = 0.0;
      
      // local variable
      momentum = 0.0;
      momentum_mass = 0.0;

      type = Extension::Barostat;

      LOG4ESPP_INFO(theLogger, "LangevinBarostat constructed");
    }

    LangevinBarostat::~LangevinBarostat(){
      LOG4ESPP_INFO(theLogger, "~LangevinBarostat");
      disconnect();
    }

    void LangevinBarostat::disconnect(){
      _runInit.disconnect();
      _befIntP.disconnect();
      _inIntP.disconnect();
      _aftIntV.disconnect();
      _aftCalcF.disconnect();
    }

    void LangevinBarostat::connect(){
      // connection to initialisation
      _runInit = integrator->runInit.connect( boost::bind(&LangevinBarostat::initialize, this));
      
      _befIntP = integrator->befIntP.connect( boost::bind(&LangevinBarostat::upd_Vp, this));
              
      _inIntP = integrator->inIntP.connect( boost::bind(&LangevinBarostat::updDisplacement, this, _1));
              
      _aftIntV = integrator->aftIntV.connect( boost::bind(&LangevinBarostat::upd_pV, this));
              
      _aftCalcF = integrator->aftCalcF.connect( boost::bind(&LangevinBarostat::updForces, this));
    }
    
    void LangevinBarostat::setGammaP(real _gammaP){
      gammaP = _gammaP;
    }
    real LangevinBarostat::getGammaP(){
      return gammaP;
    }

    void LangevinBarostat::setPressure(real _press){
      externalPressure = _press;
    }
    real LangevinBarostat::getPressure(){
      return externalPressure;
    }
    
    void LangevinBarostat::setMass(real _mass){
      mass = _mass;
    }
    real LangevinBarostat::getMass(){
      return mass;
    }
    
    // 
    void LangevinBarostat::setMassByFrequency(real freq){
      System& system = getSystemRef();
      int N = 0;
      int Nloc = system.storage->getNRealParticles();
      boost::mpi::all_reduce(*mpiWorld, Nloc, N, std::plus<int>());
      
      // TODO introduce dimension as a parameter. It will be done by introducing
      // degree of freedom as a property of System.
      real d = 3;
      
      mass = d*N*desiredTemperature / (freq*freq);
    }
    
    // it is for signals at first we modify volume then momentum
    void LangevinBarostat::upd_Vp(){
      updVolume();
      updVolumeMomentum();
    }
    // the other way around
    void LangevinBarostat::upd_pV(){
      updVolumeMomentum();
      updVolume();
    }

    void LangevinBarostat::updVolume(){
      real dt = integrator->getTimeStep();
      
      System& system = getSystemRef();
      esutil::Error err(system.comm);
      
      // The volume is scaled according to the equations V(t+1/2*dt) = V(t) + 1/2*dt*V';
      // V' = d*V*pe/W
      // in order to accelerate  1.5 * dt := 0.5dt * 3
      real volScale = 1 + dt * 1.5 * momentum_mass;
      real scale_factor = 0;
      if( volScale < 0.0){
        stringstream msg;
        msg << "Scaling coefficient is <0 (Langevin barostat). coef="<<volScale;
        err.setException( msg.str() );
        err.checkException();
      }

      scale_factor = pow( volScale, 1./3.);  // calculating the current scaling parameter
      
      system.scaleVolume( scale_factor, false);
    }
    
    // TODO should be optimized!!!
    /*
     *  TODO now it is valid for nonbonded systems because the value of degreees of freedom
     *  Nf = 3*N, N - number of particles, 3 - d-dimensional system (d=3). Thus d/Nf is 
     *  replaced by 1/N.
     */
    void LangevinBarostat::updVolumeMomentum(){
      real dt_2 = 0.5 * integrator->getTimeStep();

      // dt_2 is timestep/2. 
      System& system = getSystemRef();
      Real3D Li = system.bc -> getBoxL(); // getting the system size
      real V = Li[0] * Li[1] * Li[2];     // system volume
      real m3V = 3 * V;
      
      // get a random value and distribute the same value over all of the CPUs
      mpi::communicator communic = *system.comm;
      real rannum; 
      if (communic.rank() == 0) rannum = (*rng)()-0.5; //rannum = rng->normal();
      mpi::broadcast(communic, rannum, 0);
      
      // compute the kinetic contribution (2/3 \sum 1/2mv^2)
      // it's not efficient to use the analysis.Pressure because of double calculation of m*v*v
      real v2sum;
      real v2 = 0.0;
      CellList realCells = system.storage->getRealCells();
      for (CellListIterator cit(realCells); !cit.isDone(); ++cit) {
        const Particle& p = *cit;
        v2 += p.mass() * (p.velocity() * p.velocity());
      }
      mpi::all_reduce( communic, v2, v2sum, std::plus<real>());
      real p_kinetic = v2sum;
      
      // compute the short-range nonbonded contribution
      real rij_dot_Fij = 0.0;
      const InteractionList& srIL = system.shortRangeInteractions;
      for (size_t j = 0; j < srIL.size(); j++) {
        rij_dot_Fij += srIL[j]->computeVirial();
      }
      real p_nonbonded = rij_dot_Fij;
      // TODO optimization is needed, some terms are the same at the begin and at the end of integration
      real X = p_kinetic + p_nonbonded;
      /*
       * in order not to do double calculations X is not divided by 3V and term in the next line
       * 3V * (X - externalPressure) is replaced by (X - 3V * externalPressure)
       * local momentum derivative; in order to optimize term pref6 * v2sum could be coupled with X
       */
      real pe_deriv = (X - m3V * externalPressure) + pref6 * v2sum + pref4 * momentum + pref5 * rannum;
      
      //cout<< "externalPressure: " << externalPressure << endl;
      
      momentum += dt_2 * pe_deriv;
      momentum_mass = momentum / mass; // momentum is normalized by mass already
    }
    
    void LangevinBarostat::updForces(){
      LOG4ESPP_DEBUG(theLogger, "barostating");
      
      System& system = getSystemRef();
      CellList cells = system.storage->getRealCells();
      
      real factor = pref3  * momentum_mass;
      for(CellListIterator cit(cells); !cit.isDone(); ++cit){
        frictionBarostat(*cit, factor);
      }
    }
    
    void LangevinBarostat::updDisplacement(real& maxSqDist){
      System& system = getSystemRef();
      CellList cells = system.storage->getRealCells();
      real dt = integrator->getTimeStep();
      
      real coef = dt * momentum_mass;
      for(CellListIterator cit(cells); !cit.isDone(); ++cit){
        Particle& p = *cit;
        Real3D delta = coef * p.position();
        p.position() += delta;
        real sqDist = delta * delta;
        maxSqDist = std::max(maxSqDist, sqDist);
      }
    }

    void LangevinBarostat::frictionBarostat(Particle& p, real factor){
      p.force() += factor * p.velocity() * p.mass();
      
      LOG4ESPP_TRACE(theLogger, "new particle force = " << p.force());
    }
     
    void LangevinBarostat::initialize(){
      // calculate the prefactors
      LOG4ESPP_INFO(theLogger, "init, gammaP = " << gammaP << 
                               ", external pressure = " << externalPressure <<
                               ", fictitious mass = " << mass );

      // TODO implement Nf for the system with constarins
      /* Nf - degrees of freedom. For N particles in d-dimentional system without constrains
       * Nf = d * N
       */
      // determine the number of local particles and total particles
      System& system = getSystemRef();
      int Nsum = 0;
      int N = system.storage->getNRealParticles();
      boost::mpi::all_reduce(*mpiWorld, N, Nsum, std::plus<int>());
      
      pref6 = 1./(double)Nsum;
      
      pref3 = - (1+pref6);
      
      // pressure friction prefactor
      pref4 = -gammaP;
      
      real dt = integrator->getTimeStep();
      
      // uniform distribution prefactor. (it can be used instead of normal distribution)
      pref5 = sqrt( 8.0 * desiredTemperature * gammaP * mass / dt );
      //pref5 = sqrt( 8.0 * gammaP * mass / dt );
    }

    /****************************************************
    ** REGISTRATION WITH PYTHON
    ****************************************************/

    void LangevinBarostat::registerPython() {
      using namespace espressopp::python;

      class_<LangevinBarostat, shared_ptr<LangevinBarostat>, bases<Extension> >

        ("integrator_LangevinBarostat", init< shared_ptr<System>, shared_ptr<esutil::RNG>, real >() )

        .add_property("gammaP", &LangevinBarostat::getGammaP, &LangevinBarostat::setGammaP)
        .add_property("pressure", &LangevinBarostat::getPressure, &LangevinBarostat::setPressure)
        .add_property("mass", &LangevinBarostat::getMass, &LangevinBarostat::setMass)
      
        .def("setMassByFrequency", &LangevinBarostat::setMassByFrequency)
      
        .def("connect", &LangevinBarostat::connect)
        .def("disconnect", &LangevinBarostat::disconnect)
      ;
    }

  }
}

File: src/integrator/LangevinBarostat.py
#  Copyright (C) 2012,2013
#      Max Planck Institute for Polymer Research
#  Copyright (C) 2008,2009,2010,2011
#      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
#  
#  This file is part of ESPResSo++.
#  
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>. 


r"""
**************************************
espressopp.integrator.LangevinBarostat
**************************************

This is the barostat implementation to perform Langevin dynamics in a Hoover style extended system
according to the paper [Quigley04]_. It includes corrections of Hoover approach which were introduced
by Martyna et al [Martyna94]_.
If LangevinBarostat is defined (as a property of integrator) the integration equations will be 
modified. The volume of system :math:`V` is introduced as a dynamical variable:

.. math:: \boldsymbol{\dot{r}}_{i} = \frac{\boldsymbol{p}_{i}}{m_{i}} +
          \frac{p_{\epsilon}}{W}\boldsymbol{r}_{i}
					
.. math:: \boldsymbol{\dot{p}}_{i} = -\bigtriangledown_{\boldsymbol{r}_{i}}\Phi -
          (1+\frac{n}{N_{f}})\frac{p_{\epsilon}}{W}\boldsymbol{p}_{i} - 
          \gamma\boldsymbol{p}_{i} + \boldsymbol{R}_{i}
					
.. math:: \dot{V} = dVp_{\epsilon}/W

.. math:: \dot{p}_{\epsilon} = nV(X-P_{ext})+
          \frac{n}{N_{f}}\sum^{N}_{i=1}\frac{\boldsymbol{p}_{i}^{2}}{m_{i}} -
          \gamma_{p}p_{\epsilon} + R_{p}


where volume has a fictitious mass :math:`W` and associated momentum :math:`p_{\epsilon}`,
:math:`\gamma_{p}` - friction coefficient,
:math:`P_{ext}` - external pressure and :math:`X` - instantaneous pressure without white noise
contribution from thermostat, :math:`n` - dimension, :math:`N_{f}` - degrees of freedom (if there
are no constrains and :math:`N` is the number of particles in system :math:`N_{f}=nN`).
:math:`R_{p}` - values which are drawn from Gaussian distribution of zero mean and unit variance
scaled by

.. math:: \sqrt{\frac{2k_{B}TW\gamma_{p}}{\Delta t}}

**!IMPORTANT** Terms :math:`- \gamma\boldsymbol{p}_{i} + \boldsymbol{R}_{i}` correspond to the 
termostat. They are not included here and will not be calculated if the Langevin Thermostat is not
defined.

Example:

    >>> rng = espressopp.esutil.RNG()
    >>> langevinP = espressopp.integrator.LangevinBarostat(system, rng, desiredTemperature)
    >>> langevinP.gammaP = 0.05
    >>> langevinP.pressure = 1.0
    >>> langevinP.mass = pow(10.0, 4)
    >>> integrator.addExtension(langevinP)

**!IMPORTANT**  This barostat is supposed to be run in a couple with thermostat in order 
to simulate the *npt* ensamble, because the term :math:`R_{p}` needs the temperature as a 
parameter. 

Definition:

    In order to define the Langevin-Hoover barostat
    
    >>> langevinP = espressopp.integrator.LangevinBarostat(system, rng, desiredTemperature)
    
    one should have the System_ and RNG_ defined and know the desired temperature.

.. _System: espressopp.System.html
.. _RNG:

Properties:

*   *langevinP.gammaP*

    The property 'gammaP' defines the friction coefficient :math:`\gamma_{p}`.

*   *langevinP.pressure*
    
    The property 'pressure' defines the external pressure :math:`P_{ext}`.
    
*   *langevinP.mass*

    The property 'mass' defines the fictitious mass :math:`W`.

Methods:

*   *setMassByFrequency( frequency )*

    Set the proper *langevinP.mass* using expression :math:`W=dNk_{b}T/\omega_{b}^{2}`, where
    frequency, :math:`\omega_{b}`, is the frequency of required volume fluctuations. The value of
    :math:`\omega_{b}` should be less then the lowest frequency which appears in the NVT temperature
    spectrum [Quigley04]_ in order to match the canonical distribution. :math:`d` - dimensions,
    :math:`N` - number of particles, :math:`k_{b}` - Boltzmann constant, :math:`T` - desired
    temperature.
    
**NOTE** The *langevinP.mass* can be set both directly and using the 
(*setMassByFrequency( frequency )*) 

Adding to the integration:
    
    >>> integrator.addExtension(langevinP)
    
    It will define Langevin-Hoover barostat as a property of integrator.
    
One more example:

    >>> rngBaro = espressopp.esutil.RNG()
    >>> lP = espressopp.integrator.LangevinBarostat(system, rngBaro, desiredTemperature)
    >>> lP.gammaP = .5
    >>> lP.pressure = 1.0
    >>> lP.mass = pow(10.0, 5)
    >>> integrator.addExtension(lP)

Canceling the barostat:
    
    If one do not need the pressure regulation in system anymore or need to switch
    the ensamble or whatever :)

    >>> # define barostat with parameters
    >>> rngBaro = espressopp.esutil.RNG()
    >>> lP = espressopp.integrator.LangevinBarostat(system, rngBaro, desiredTemperature)
    >>> lP.gammaP = .5
    >>> lP.pressure = 1.0
    >>> lP.mass = pow(10.0, 5)
    >>> integrator.langevinBarostat = lP
    >>> ...
    >>> # some runs
    >>> ...
    >>> # disconnect barostat
    >>> langevinBarostat.disconnect()
    >>> # the next runs will not include the modification of integration equations

    
  Connecting the barostat back after the disconnection
  
    >>> langevinBarostat.connect()

References:

.. [Quigley04] D. Quigley, M.I.J. Probert, *J. Chem. Phys.*, 120, **2004**, p. 11432

.. [Martyna94] G. Martyna, D. Tobias, M. Klein, *J. Chem. Phys.*, 101, **1994**, p. 4177


.. function:: espressopp.integrator.LangevinBarostat(system, rng, temperature)

		:param system: 
		:param rng: 
		:param temperature: 
		:type system: 
		:type rng: 
		:type temperature: 
"""


from espressopp.esutil import cxxinit
from espressopp import pmi

from espressopp.integrator.Extension import *
from _espressopp import integrator_LangevinBarostat

class LangevinBarostatLocal(ExtensionLocal, integrator_LangevinBarostat):
  def __init__(self, system, rng, temperature):

    if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
      cxxinit(self, integrator_LangevinBarostat, system, rng, temperature)

if pmi.isController :
  class LangevinBarostat(Extension):
    __metaclass__ = pmi.Proxy
    pmiproxydefs = dict(
      cls =  'espressopp.integrator.LangevinBarostatLocal',
      pmiproperty = [ 'gammaP', 'pressure', 'mass' ],
      pmicall = [ "setMassByFrequency" ]
    )
